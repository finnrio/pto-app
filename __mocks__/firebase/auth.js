// __mocks__/firebase/auth.js
export const getAuth = jest.fn(() => {
  return {
    currentUser: { uid: 'testUid' },
    signInWithEmailAndPassword: jest.fn(),
    signOut: jest.fn(),
  };
});
