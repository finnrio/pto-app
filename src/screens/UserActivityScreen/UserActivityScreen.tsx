import { Alert, View } from "react-native";
import React, { useContext, useState } from "react";
import { List } from "react-native-paper";
import { useFocusEffect } from "@react-navigation/native";
import styles from "./styles";
import GetUserData from "../../firebase/operations/GetUserData";
import GetPTOByStatus from "../../firebase/operations/GetPTOByStatus";
import { UserContext } from "../../context/UserContext";
import DeletePTORequest from "../../firebase/operations/DeletePTORequest";

export default function UserActivityScreen() {
  const currentUser = useContext(UserContext);
  const [pendingPTO, setPendingPTO] = useState<any[]>([]);
  const [approvedPTO, setApprovedPTO] = useState<any[]>([]);
  const [deniedPTO, setDeniedPTO] = useState<any[]>([]);
  const [refresh, setRefresh] = useState(false);

  useFocusEffect(
    React.useCallback(() => {
      GetPTOByStatus(currentUser!.uid, "Pending").then((data) => {
        setPendingPTO(data);
      });
      GetPTOByStatus(currentUser!.uid, "Approved").then((data) => {
        setApprovedPTO(data);
      });
      GetPTOByStatus(currentUser!.uid, "Denied").then((data) => {
        setDeniedPTO(data);
      });
    }, [refresh]),
  );

  async function queryAlert(data: any) {
    const user = await GetUserData(data.user_id);
    Alert.alert(
      "PTO Request",
      `Employee: ${user.first_name} ${user.surname}\nReason: ${data.reason}\nStarting: ${data.start_date}\nEnding: ${data.end_date}\nHours Requested: ${data.hours}`,
      [
        {
          text: "Delete PTO Request",
          onPress: async () => {
            if (data.start_date < new Date().toISOString().split("T")[0]) {
              Alert.alert("Error", "Cannot delete past PTO requests");
              return;
            }
            console.log(
              "User ",
              data.user_id,
              " deleted pto request: ",
              data.id,
            );
            await DeletePTORequest(data.user_id, data.id).then(() =>
              setRefresh(!refresh),
            );
          },
          style: "default",
        },
        {
          text: "Cancel",
          style: "cancel",
        },
      ],
      { cancelable: true },
    );
  }

  function createListItem(data: any) {
    return (
      <List.Item
        title={`${data.start_date} to ${data.end_date}`}
        key={data.id}
        onPress={() => queryAlert(data)}
      />
    );
  }

  return (
    <View style={styles.container}>
      <List.AccordionGroup>
        <List.Accordion title="Pending" id="1">
          {pendingPTO.map((data) => createListItem(data))}
        </List.Accordion>
        <List.Accordion title="Approved" id="2">
          {approvedPTO.map((data) => createListItem(data))}
        </List.Accordion>
        <List.Accordion title="Denied" id="3">
          {deniedPTO.map((data) => createListItem(data))}
        </List.Accordion>
      </List.AccordionGroup>
    </View>
  );
}
