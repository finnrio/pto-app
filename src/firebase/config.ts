import { initializeApp } from 'firebase/app';
import { getAuth } from "firebase/auth";
import { getFirestore } from "firebase/firestore";

// Optionally import the services that you want to use
// import {...} from "firebase/auth";
// import {...} from "firebase/database";
// import {...} from "firebase/firestore";
// import {...} from "firebase/functions";
// import {...} from "firebase/storage";

// Initialize Firebase
const firebaseConfig = {
  apiKey: 'AIzaSyBEjW_646bH_Q4ECQD_g7nXhM_WTDQQZqw',
  authDomain: 'pto-app-30271.firebaseapp.com',
  databaseURL: 'https://pto-app-30271.firebaseio.com',
  projectId: 'pto-app-30271',
  storageBucket: 'pto-app-30271.appspot.com',
  messagingSenderId: '440515616929',
  appId: '1:440515616929:ios:1be65de5cfa7a0048210e0',
  measurementId: '1:440515616929:ios:1be65de5cfa7a0048210e0',
};

export const FIREBASE_APP = initializeApp(firebaseConfig);
// For more information on how to access Firebase in your project,
// see the Firebase documentation: https://firebase.google.com/docs/web/setup#access-firebase

export const FIREBASE_AUTH = getAuth(FIREBASE_APP);
export const FIRESTORE_DB = getFirestore(FIREBASE_APP);