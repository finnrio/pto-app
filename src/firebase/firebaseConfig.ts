import { initializeApp } from "firebase/app";
import { connectAuthEmulator, getAuth } from "firebase/auth";
import { connectFirestoreEmulator, getFirestore } from "firebase/firestore";

// Initialize Firebase
const firebaseConfig = {
  apiKey: "AIzaSyBEjW_646bH_Q4ECQD_g7nXhM_WTDQQZqw",
  authDomain: "pto-app-30271.firebaseapp.com",
  databaseURL: "https://pto-app-30271.firebaseio.com",
  projectId: "pto-app-30271",
  storageBucket: "pto-app-30271.appspot.com",
  messagingSenderId: "440515616929",
  appId: "1:440515616929:ios:1be65de5cfa7a0048210e0",
  measurementId: "1:440515616929:ios:1be65de5cfa7a0048210e0",
};

const FIREBASE_APP = initializeApp(firebaseConfig);

const FIREBASE_AUTH = getAuth(FIREBASE_APP);
const FIRESTORE_DB = getFirestore(FIREBASE_APP);

if (process.env.NODE_ENV === "test") {
  console.log("connecting to emulator");
  connectAuthEmulator(FIREBASE_AUTH, "http://127.0.0.1:9099");
  connectFirestoreEmulator(FIRESTORE_DB, "http://127.0.0.1", 8080);
}

export { FIREBASE_APP, FIREBASE_AUTH, FIRESTORE_DB };
