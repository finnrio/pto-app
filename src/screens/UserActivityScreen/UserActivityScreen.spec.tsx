import React from "react";
import { render } from "@testing-library/react-native";

import { User } from "firebase/auth";
import { createMock } from "@golevelup/ts-jest";
import { NavigationContainer } from "@react-navigation/native";
import { UserContext } from "../../context/UserContext";
import UserActivityScreen from "./UserActivityScreen";

const mockAuthUser: User = createMock<User>({ uid: "mock_uid" });

describe("UserActivityScreen", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("renders correctly", () => {
    const { toJSON } = render(
      <NavigationContainer>
        <UserContext.Provider value={mockAuthUser}>
          <UserActivityScreen />
        </UserContext.Provider>
      </NavigationContainer>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
