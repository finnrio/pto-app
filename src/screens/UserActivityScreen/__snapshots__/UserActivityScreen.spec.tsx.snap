// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserActivityScreen renders correctly 1`] = `
<View
  style={
    {
      "alignContent": "center",
      "flex": 1,
      "width": "100%",
    }
  }
>
  <View>
    <View
      style={
        {
          "backgroundColor": "rgba(255, 251, 254, 1)",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": false,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            [
              {
                "paddingRight": 24,
                "paddingVertical": 8,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            {
              "flexDirection": "row",
              "marginVertical": 6,
            }
          }
        >
          <View
            style={
              [
                {
                  "paddingLeft": 16,
                },
                {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 16,
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Pending
            </Text>
          </View>
          <View
            style={
              [
                {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                undefined,
              ]
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        {
          "backgroundColor": "rgba(255, 251, 254, 1)",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": false,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            [
              {
                "paddingRight": 24,
                "paddingVertical": 8,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            {
              "flexDirection": "row",
              "marginVertical": 6,
            }
          }
        >
          <View
            style={
              [
                {
                  "paddingLeft": 16,
                },
                {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 16,
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Approved
            </Text>
          </View>
          <View
            style={
              [
                {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                undefined,
              ]
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        {
          "backgroundColor": "rgba(255, 251, 254, 1)",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": false,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            [
              {
                "paddingRight": 24,
                "paddingVertical": 8,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            {
              "flexDirection": "row",
              "marginVertical": 6,
            }
          }
        >
          <View
            style={
              [
                {
                  "paddingLeft": 16,
                },
                {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 16,
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Denied
            </Text>
          </View>
          <View
            style={
              [
                {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                undefined,
              ]
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
